File: app/src/main/java/com/example/agriwaste/AgriWasteDataHub.java
package com.example.agriwaste;

import com.example.model.AgriWasteItem;
import com.example.model.AgriWasteOrder;
import com.example.model.AgriWastePurchaseRequest;
import com.example.model.WasteTypeMetadata;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Thread-safe Central Data Hub for the Agri Waste Marketplace.
 * Enforces strict role boundaries:
 * - Farmer: Only Lists, Edits, Deletes, Views Own Listings, and Accepts/Rejects buyer requests. Cannot buy.
 * - Buyer: Browses, Searches, Filters, Views Details, and Sends Purchase Requests / Orders.
 */
public class AgriWasteDataHub {

    private static volatile AgriWasteDataHub instance;

    private final CopyOnWriteArrayList<AgriWasteItem> listings = new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<AgriWasteOrder> orders = new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<AgriWastePurchaseRequest> purchaseRequests = new CopyOnWriteArrayList<>();
    private final List<WasteTypeMetadata> wasteTypes = new ArrayList<>();

    private AgriWasteDataHub() {
        initWasteTypes();
        seedInitialListings();
    }

    public static AgriWasteDataHub getInstance() {
        if (instance == null) {
            synchronized (AgriWasteDataHub.class) {
                if (instance == null) {
                    instance = new AgriWasteDataHub();
                }
            }
        }
        return instance;
    }

    private void initWasteTypes() {
        wasteTypes.add(new WasteTypeMetadata(
                "Wheat Straw", "🌾", "Straw", "kg", "₹/kg",
                Arrays.asList("Animal fodder & cattle feed", "Mushroom cultivation substrate", "Organic vermicompost", "Biomass briquettes"),
                "High demand from dairy clusters & biomass plants"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Rice Straw", "🌾", "Straw", "ton", "₹/ton",
                Arrays.asList("Cattle bedding & feed", "Paddy mushroom growing", "Bio-energy power plants", "Eco-packaging boards"),
                "Crucial for stubble burning prevention & bio-energy"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Sugarcane Residue", "🎋", "Sugarcane Residue", "ton", "₹/ton",
                Arrays.asList("Soil mulching & moisture conservation", "Sugar mill cogeneration boilers", "Compost & vermiwash", "Eco-paper pulp"),
                "Very high calorific value for industrial boilers"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Maize Stalk", "🌽", "Crop Stalks", "ton", "₹/ton",
                Arrays.asList("Nutritious silage & green fodder", "Industrial boiler fuel", "Field mulch cover", "Bio-char conversion"),
                "Essential for dairy silage in summer"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Cotton Stalk", "🌱", "Crop Stalks", "ton", "₹/ton",
                Arrays.asList("High-density biomass briquettes", "Particle board manufacturing", "Industrial boiler fuel", "Activated carbon"),
                "Preferred wood substitute for biomass fuel"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Soybean Residue", "🫘", "Other Crop Residue", "quintal", "₹/quintal",
                Arrays.asList("High protein milch cattle fodder", "Soil nitrogen fixation compost", "Animal bedding"),
                "Rapidly purchased by local dairy farmers"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Groundnut Shell", "🥜", "Husk / Shell", "quintal", "₹/quintal",
                Arrays.asList("Biofuel pellets & briquettes", "Poultry farm bedding & litter", "Activated carbon filtration"),
                "High heating value for heating plants"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Coconut Husk", "🥥", "Husk / Shell", "bundle", "₹/bundle",
                Arrays.asList("Coir fiber rope & geotextiles", "Coco-peat for plant nurseries & hydroponics", "Soil water-retention mulching"),
                "Huge export & domestic nursery demand"
        ));
        wasteTypes.add(new WasteTypeMetadata(
                "Other", "♻️", "Other Crop Residue", "kg", "₹/kg",
                Arrays.asList("General farm composting", "Bio-char generation", "Soil mulching", "Local biomass fuel"),
                "Eco-friendly zero-waste farm management"
        ));
    }

    private void seedInitialListings() {
        listings.add(new AgriWasteItem(
                "AW-LST-101", "FARMER_MH_01", "Rameshwar Patil", "+91 98220 12345",
                "Wheat Straw", "Clean Golden Wheat Straw", "Straw", 15.0, 15.0,
                "ton", 1800.0, "₹/ton", "30 Aug 2026", "Narayangaon", "Junnar", "Pune",
                3.5, "Dry, well packed golden wheat straw. Ideal for biomass pellets & animal feed.",
                "🌾", "Both Supported", "Available", "26 Aug 2026", "26 Aug 2026"
        ));
        listings.add(new AgriWasteItem(
                "AW-LST-102", "FARMER_MH_01", "Rameshwar Patil", "+91 98220 12345",
                "Sugarcane Residue", "Shredded Sugarcane Bagasse & Tops", "Sugarcane Residue", 25.0, 25.0,
                "ton", 1400.0, "₹/ton", "05 Sept 2026", "Narayangaon", "Junnar", "Pune",
                4.2, "High calorific value sugarcane dry residue and bagasse for cogeneration plants.",
                "🎋", "Delivery Partner", "Available", "26 Aug 2026", "26 Aug 2026"
        ));
        listings.add(new AgriWasteItem(
                "AW-LST-103", "f_anand", "Anand Kadam", "+91 94225 99887",
                "Rice Straw", "High Quality Paddy Straw Bundles", "Straw", 10.0, 10.0,
                "ton", 1600.0, "₹/ton", "02 Sept 2026", "Ozar", "Junnar", "Pune",
                6.0, "Well tied paddy straw bales for mushroom cultivation and dairy bedding.",
                "🌾", "Both Supported", "Available", "25 Aug 2026", "25 Aug 2026"
        ));

        // Sample Buyer Purchase Request to Farmer
        purchaseRequests.add(new AgriWastePurchaseRequest(
                "REQ-AW-501", "AW-LST-101", "FARMER_MH_01", "Rameshwar Patil",
                "BUYER_01", "GreenBio Energy Solutions", "+91 98900 77112", "Biomass Plant",
                "Clean Golden Wheat Straw", 10.0, "ton", 1750.0, "₹/ton", 17500.0,
                "02 Sept 2026", "MIDC Chakan Industrial Area, Pune", "Pending",
                "26 Aug 2026", "Need bulk dispatch for our biomass briquetting unit."
        ));
    }

    public List<WasteTypeMetadata> getWasteTypes() {
        return new ArrayList<>(wasteTypes);
    }

    // --- Farmer Scope: Listing Management ---
    public List<AgriWasteItem> getFarmerListings(String farmerId) {
        List<AgriWasteItem> result = new ArrayList<>();
        for (AgriWasteItem item : listings) {
            if (item.getFarmerId().equals(farmerId)) {
                result.add(item);
            }
        }
        return result;
    }

    public synchronized void addListing(AgriWasteItem item) {
        if (item != null) {
            listings.add(0, item);
        }
    }

    public synchronized boolean updateListing(AgriWasteItem updatedItem) {
        if (updatedItem == null) return false;
        for (int i = 0; i < listings.size(); i++) {
            if (listings.get(i).getId().equals(updatedItem.getId())) {
                listings.set(i, updatedItem);
                return true;
            }
        }
        return false;
    }

    public synchronized boolean deleteListing(String listingId) {
        return listings.removeIf(item -> item.getId().equals(listingId));
    }

    // --- Buyer Scope: Browsing & Ordering ---
    public List<AgriWasteItem> getAllAvailableListings() {
        List<AgriWasteItem> result = new ArrayList<>();
        for (AgriWasteItem item : listings) {
            if ("Available".equalsIgnoreCase(item.getStatus())) {
                result.add(item);
            }
        }
        return result;
    }

    public AgriWasteItem getListingById(String id) {
        for (AgriWasteItem item : listings) {
            if (item.getId().equals(id)) {
                return item;
            }
        }
        return null;
    }

    // --- Purchase Requests Workflow ---
    public List<AgriWastePurchaseRequest> getRequestsForFarmer(String farmerId) {
        List<AgriWastePurchaseRequest> result = new ArrayList<>();
        for (AgriWastePurchaseRequest req : purchaseRequests) {
            if (req.getFarmerId().equals(farmerId)) {
                result.add(req);
            }
        }
        return result;
    }

    public synchronized void submitPurchaseRequest(AgriWastePurchaseRequest request) {
        if (request != null) {
            purchaseRequests.add(0, request);
        }
    }

    public synchronized boolean respondToPurchaseRequest(String requestId, boolean accept) {
        for (AgriWastePurchaseRequest req : purchaseRequests) {
            if (req.getId().equals(requestId)) {
                req.setStatus(accept ? "Accepted" : "Rejected");
                if (accept) {
                    // Create an order in the pipeline
                    String orderId = "ORD-AW-" + System.currentTimeMillis();
                    AgriWasteOrder order = new AgriWasteOrder(
                            orderId, req.getListingId(), req.getFarmerId(), req.getFarmerName(), "",
                            req.getBuyerId(), req.getBuyerName(), req.getBuyerPhone(), req.getBuyerType(),
                            req.getWasteName(), "Straw", req.getRequestedQuantity(), req.getUnit(),
                            req.getOfferedPrice(), req.getPriceUnit(), req.getTotalAmount(),
                            "Delivery Partner", req.getDeliveryAddress(), "Narayangaon", "Junnar", "Pune",
                            null, null, null, "Accepted",
                            new SimpleDateFormat("dd MMM yyyy", Locale.ENGLISH).format(new Date()),
                            req.getProposedPickupDate(), null, req.getNotes()
                    );
                    orders.add(0, order);
                }
                return true;
            }
        }
        return false;
    }

    // --- Orders Management ---
    public List<AgriWasteOrder> getAllOrders() {
        return new ArrayList<>(orders);
    }

    public List<AgriWasteOrder> getOrdersForFarmer(String farmerId) {
        List<AgriWasteOrder> result = new ArrayList<>();
        for (AgriWasteOrder order : orders) {
            if (order.getFarmerId().equals(farmerId)) {
                result.add(order);
            }
        }
        return result;
    }

    public AgriWasteOrder getOrderById(String orderId) {
        if (orderId == null) return null;
        for (AgriWasteOrder order : orders) {
            if (order.getId().equals(orderId)) {
                return order;
            }
        }
        return null;
    }

    public synchronized boolean updateOrderStatus(String orderId, String newStatus) {
        for (AgriWasteOrder order : orders) {
            if (order.getId().equals(orderId)) {
                order.setStatus(newStatus);
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/java/com/example/AgriWasteDataHub.kt
package com.example

import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.snapshots.SnapshotStateList

// ------------------ DATA MODELS ------------------

data class WasteTypeMetadata(
    val name: String,
    val emoji: String,
    val category: String,
    val defaultUnit: String,
    val defaultPriceUnit: String,
    val typicalUses: List<String>,
    val industrialDemand: String
)

val AGRI_WASTE_TYPES = listOf(
    WasteTypeMetadata(
        name = "Wheat Straw",
        emoji = "🌾",
        category = "Straw",
        defaultUnit = "kg",
        defaultPriceUnit = "₹/kg",
        typicalUses = listOf("Animal fodder & cattle feed", "Mushroom cultivation substrate", "Organic vermicompost", "Biomass briquette manufacturing"),
        industrialDemand = "High demand from dairy clusters & biomass plants"
    ),
    WasteTypeMetadata(
        name = "Rice Straw",
        emoji = "🌾",
        category = "Straw",
        defaultUnit = "ton",
        defaultPriceUnit = "₹/ton",
        typicalUses = listOf("Cattle bedding & feed", "Paddy straw mushroom growing", "Bio-energy power plants", "Eco-packaging boards"),
        industrialDemand = "Crucial for stubble burning prevention & bio-energy"
    ),
    WasteTypeMetadata(
        name = "Sugarcane Residue",
        emoji = "🎋",
        category = "Sugarcane Residue",
        defaultUnit = "ton",
        defaultPriceUnit = "₹/ton",
        typicalUses = listOf("Soil mulching & moisture conservation", "Sugar mill cogeneration boilers", "Compost & vermiwash", "Eco-paper pulp"),
        industrialDemand = "Very high calorific value for industrial boilers"
    ),
    WasteTypeMetadata(
        name = "Maize Stalk",
        emoji = "🌽",
        category = "Crop Stalks",
        defaultUnit = "ton",
        defaultPriceUnit = "₹/ton",
        typicalUses = listOf("Nutritious silage & green fodder", "Industrial boiler fuel", "Field mulch cover", "Bio-char conversion"),
        industrialDemand = "Essential for dairy silage in summer"
    ),
    WasteTypeMetadata(
        name = "Cotton Stalk",
        emoji = "🌱",
        category = "Crop Stalks",
        defaultUnit = "ton",
        defaultPriceUnit = "₹/ton",
        typicalUses = listOf("High-density biomass briquettes", "Particle board manufacturing", "Industrial boiler fuel", "Activated carbon"),
        industrialDemand = "Preferred wood substitute for biomass fuel"
    ),
    WasteTypeMetadata(
        name = "Soybean Residue",
        emoji = "🫘",
        category = "Other Crop Residue",
        defaultUnit = "quintal",
        defaultPriceUnit = "₹/quintal",
        typicalUses = listOf("High protein milch cattle fodder", "Soil nitrogen fixation compost", "Animal bedding"),
        industrialDemand = "Rapidly purchased by local dairy farmers"
    ),
    WasteTypeMetadata(
        name = "Groundnut Shell",
        emoji = "🥜",
        category = "Husk / Shell",
        defaultUnit = "quintal",
        defaultPriceUnit = "₹/quintal",
        typicalUses = listOf("Biofuel pellets & briquettes", "Poultry farm bedding & litter", "Activated carbon filtration"),
        industrialDemand = "High heating value for heating plants"
    ),
    WasteTypeMetadata(
        name = "Coconut Husk",
        emoji = "🥥",
        category = "Husk / Shell",
        defaultUnit = "bundle",
        defaultPriceUnit = "₹/bundle",
        typicalUses = listOf("Coir fiber rope & geotextiles", "Coco-peat for plant nurseries & hydroponics", "Soil water-retention mulching"),
        industrialDemand = "Huge export & domestic nursery demand"
    ),
    WasteTypeMetadata(
        name = "Other",
        emoji = "♻️",
        category = "Other Crop Residue",
        defaultUnit = "kg",
        defaultPriceUnit = "₹/kg",
        typicalUses = listOf("General farm composting", "Bio-char generation", "Soil mulching", "Local biomass fuel"),
        industrialDemand = "Eco-friendly zero-waste farm management"
    )
)

data class AgriWasteItem(
    val id: String,
    val farmerId: String,
    val farmerName: String,
    val farmerPhone: String,
    val wasteType: String, // from AGRI_WASTE_TYPES names
    val wasteName: String,
    val category: String, // "Straw", "Crop Stalks", "Sugarcane Residue", "Husk / Shell", "Other Crop Residue"
    var quantity: Double,
    val initialQuantity: Double,
    val unit: String, // "kg", "quintal", "ton", "bundle", "bags"
    val price: Double,
    val priceUnit: String, // "₹/kg", "₹/quintal", "₹/ton", "₹/bundle", "₹/bag"
    val availableDate: String,
    val village: String,
    val taluka: String,
    val district: String,
    val distanceKm: Double = 4.5,
    val description: String,
    val imageEmoji: String = "🌾",
    val pickupPreference: String = "Both Supported", // "Buyer Pickup", "Delivery Partner", "Both Supported"
    var status: String = "Available", // "Available", "Sold Out", "Cancelled"
    val createdAt: String = "26 Aug 2026",
    val updatedAt: String = "26 Aug 2026"
)

data class AgriWasteOrder(
    val id: String,
    val wasteId: String,
    val farmerId: String,
    val farmerName: String,
    val farmerPhone: String,
    val buyerId: String,
    val buyerName: String,
    val buyerPhone: String,
    val buyerType: String = "Biomass Buyer", // "Dairy Farm", "Biomass Plant", "Compost Unit", "Nursery", "Trader"
    val wasteName: String,
    val wasteType: String,
    val quantity: Double,
    val unit: String,
    val agreedPrice: Double,
    val priceUnit: String,
    val totalAmount: Double,
    val pickupMethod: String, // "Buyer Pickup", "Delivery Partner"
    val deliveryAddress: String,
    val village: String,
    val taluka: String,
    val district: String,
    var deliveryPartnerId: String? = null,
    var deliveryPartnerName: String? = null,
    var deliveryPartnerPhone: String? = null,
    var status: String = "Waiting for Farmer", // "Waiting for Farmer", "Accepted", "Pickup Scheduled", "Picked Up", "Out for Delivery", "Delivered", "Completed", "Rejected"
    val orderDate: String,
    val pickupDate: String,
    var completedDate: String? = null,
    val notes: String = ""
)

data class AgriWasteNotificationItem(
    val id: String,
    val title: String,
    val message: String,
    val timestamp: String,
    val type: String, // "Request", "Accepted", "Delivery", "Completed"
    val recipientRole: String, // "farmer", "buyer", "delivery"
    var isRead: Boolean = false
)

// ------------------ SINGLETON REPOSITORY ------------------

object AgriWasteDataHub {

    val listings: SnapshotStateList<AgriWasteItem> = mutableStateListOf()

    val orders: SnapshotStateList<AgriWasteOrder> = mutableStateListOf()

    val notifications: SnapshotStateList<AgriWasteNotificationItem> = mutableStateListOf()

    // Helper functions

    fun addListing(item: AgriWasteItem) {
        listings.add(0, item)
        notifications.add(0, AgriWasteNotificationItem(
            id = "notif_" + System.currentTimeMillis(),
            title = "Waste Listing Published 🌾",
            message = "Your listing for ${item.wasteName} (${item.quantity} ${item.unit}) is now visible to buyers.",
            timestamp = "Just now",
            type = "Accepted",
            recipientRole = "farmer"
        ))
    }

    fun updateListing(updated: AgriWasteItem) {
        val idx = listings.indexOfFirst { it.id == updated.id }
        if (idx != -1) {
            listings[idx] = updated
        }
    }

    fun deleteListing(id: String) {
        listings.removeAll { it.id == id }
    }

    fun placeOrder(order: AgriWasteOrder): String {
        orders.add(0, order)
        // Notify farmer
        notifications.add(0, AgriWasteNotificationItem(
            id = "notif_" + System.currentTimeMillis(),
            title = "♻️ Waste Purchase Request",
            message = "${order.buyerName} requested ${order.quantity} ${order.unit} of ${order.wasteName} (${order.pickupMethod}).",
            timestamp = "Just now",
            type = "Request",
            recipientRole = "farmer"
        ))
        return order.id
    }

    fun acceptOrder(orderId: String) {
        val idx = orders.indexOfFirst { it.id == orderId }
        if (idx != -1) {
            val ord = orders[idx]
            val nextStatus = if (ord.pickupMethod == "Buyer Pickup") "Accepted" else "Pickup Scheduled"
            orders[idx] = ord.copy(status = nextStatus)

            // Deduct available quantity from listing
            val listIdx = listings.indexOfFirst { it.id == ord.wasteId }
            if (listIdx != -1) {
                val current = listings[listIdx]
                val remaining = (current.quantity - ord.quantity).coerceAtLeast(0.0)
                val newStatus = if (remaining <= 0.0) "Sold Out" else current.status
                listings[listIdx] = current.copy(quantity = remaining, status = newStatus)
            }

            // Notify buyer
            notifications.add(0, AgriWasteNotificationItem(
                id = "notif_" + System.currentTimeMillis(),
                title = "Order Confirmed by Farmer ✅",
                message = "Farmer ${ord.farmerName} confirmed your order for ${ord.quantity} ${ord.unit} ${ord.wasteName}.",
                timestamp = "Just now",
                type = "Accepted",
                recipientRole = "buyer"
            ))
        }
    }

    fun rejectOrder(orderId: String) {
        val idx = orders.indexOfFirst { it.id == orderId }
        if (idx != -1) {
            val ord = orders[idx]
            orders[idx] = ord.copy(status = "Rejected")

            notifications.add(0, AgriWasteNotificationItem(
                id = "notif_" + System.currentTimeMillis(),
                title = "Order Request Declined ❌",
                message = "Farmer ${ord.farmerName} could not fulfill the request for ${ord.wasteName}.",
                timestamp = "Just now",
                type = "Request",
                recipientRole = "buyer"
            ))
        }
    }

    fun updateOrderStatus(orderId: String, newStatus: String) {
        val idx = orders.indexOfFirst { it.id == orderId }
        if (idx != -1) {
            val ord = orders[idx]
            val completedDate = if (newStatus == "Completed" || newStatus == "Delivered") "Today" else ord.completedDate
            orders[idx] = ord.copy(status = newStatus, completedDate = completedDate)

            // If order completed and was buyer pickup, mark completed
            notifications.add(0, AgriWasteNotificationItem(
                id = "notif_" + System.currentTimeMillis(),
                title = "Order Status Updated: $newStatus",
                message = "Order #${ord.id} (${ord.wasteName}) is now marked as $newStatus.",
                timestamp = "Just now",
                type = "Delivery",
                recipientRole = "buyer"
            ))
        }
    }
}
